Tests/GRPCCoreTests/Transport/InProcessClientTransportTests.swift

```diff
@@ -0,0 +1,330 @@
+      server: .init(),
+      executionConfigurations: .init(defaultConfiguration: .init(retryPolicy: retryPolicy))
+    )
+
+    client.close()
+
+    await XCTAssertThrowsRPCErrorAsync {
+      try await client.connect(lazily: false)
+    } errorHandler: { error in
+      XCTAssertEqual(error.code, .failedPrecondition)
+    }
+  }
+
+  func testConnectWhenClosedAfterCancellation() async throws {
+    let retryPolicy = RetryPolicy(
+      maximumAttempts: 10,
+      initialBackoff: .seconds(1),
+      maximumBackoff: .seconds(1),
+      backoffMultiplier: 1.0,
+      retryableStatusCodes: [.unavailable]
+    )
+    let client = InProcessClientTransport(
+      server: .init(),
+      executionConfigurations: .init(defaultConfiguration: .init(retryPolicy: retryPolicy))
+    )
+
+    try await withThrowingTaskGroup(of: Void.self) { group in
+      group.addTask {
+        try await client.connect(lazily: false)
+      }
+      group.addTask {
+        try await Task.sleep(for: .milliseconds(100))
+      }
+
+      try await group.next()
+      group.cancelAll()
+
+      await XCTAssertThrowsRPCErrorAsync {
+        try await client.connect(lazily: false)
+      } errorHandler: { error in
+        XCTAssertEqual(error.code, .failedPrecondition)
+      }
+    }
+  }
+
+  func testCloseWhenUnconnected() {
+    let retryPolicy = RetryPolicy(
+      maximumAttempts: 10,
+      initialBackoff: .seconds(1),
+      maximumBackoff: .seconds(1),
+      backoffMultiplier: 1.0,
+      retryableStatusCodes: [.unavailable]
+    )
+    let client = InProcessClientTransport(
+      server: .init(),
+      executionConfigurations: .init(defaultConfiguration: .init(retryPolicy: retryPolicy))
+    )
+
+    XCTAssertNoThrow(client.close())
+  }
+
+  func testCloseWhenClosed() {
+    func testCloseWhenUnconnected() {
+      let retryPolicy = RetryPolicy(
+        maximumAttempts: 10,
+        initialBackoff: .seconds(1),
+        maximumBackoff: .seconds(1),
+        backoffMultiplier: 1.0,
+        retryableStatusCodes: [.unavailable]
+      )
+      let client = InProcessClientTransport(
+        server: .init(),
+        executionConfigurations: .init(defaultConfiguration: .init(retryPolicy: retryPolicy))
+      )
+
+      client.close()
+      XCTAssertNoThrow(client.close())
+    }
+  }
+
+  func testConnectSuccessfullyAndThenClose() async throws {
+    let retryPolicy = RetryPolicy(
+      maximumAttempts: 10,
+      initialBackoff: .seconds(1),
+      maximumBackoff: .seconds(1),
+      backoffMultiplier: 1.0,
+      retryableStatusCodes: [.unavailable]
+    )
+    let client = InProcessClientTransport(
+      server: .init(),
+      executionConfigurations: .init(defaultConfiguration: .init(retryPolicy: retryPolicy))
+    )
+
+    try await withThrowingTaskGroup(of: Void.self) { group in
+      group.addTask {
+        try await client.connect(lazily: false)
+      }
+      group.addTask {
+        try await Task.sleep(for: .milliseconds(100))
+      }
```

Comment:
Why do we need this sleep? Can't we just connect and then close?

Reply:
Connect won't return until `closed` is called so we must do this in separate tasks. The sleep is there to make sure that there isn't a race and the state change in `connect` happens before we call `close`

Reply:
🤦‍♂️